File: Company/Amazon/Random/Container with most water.java
class Solution {

	public int maxArea(int[] height) {

		int left = 0;
		int right = height.length - 1;

		int maxArea = 0;

		while (left < right) {

			int leftHeight = height[left];
			int rightHeight = height[right];

			int currentArea = getArea(leftHeight, rightHeight, right - left);
			maxArea = Math.max(maxArea, currentArea);

			if (leftHeight < rightHeight) {
				left ++;
			} else {
				right --;
			}
		}

		return maxArea;


	}



    public int maxArea(int[] height) {
        
    	int maxArea = 0;

    	for (int i = 0; i < height.length; i ++) {

    		int leftHeight = height[i];

    		for (int j = i + 1; j < height.length; j ++) {

    			int rightHeight = height[j];
    			int currentArea = getArea(leftHeight, rightHeight, j - i);
    			maxArea = Math.max(maxArea, currentArea);
    		}


    	}

    	return maxArea;



    }

    public int getArea(int left, int right, int base) {
    	return Math.min(left, right) * base;
    }


}

File: Two Pointer/Container With Most Water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        
        max_area = 0

        left = 0
        right = len(height) - 1

        while left < right:

            container_length = right - left 
            left_wall_height = height[left]
            right_wall_height = height[right]

            container_area = container_length * min(left_wall_height, right_wall_height)

            max_area = max(max_area, container_area)
            
            # Container area is limited by whichever is smaller of the two sides - even if we find a taller 
            # wall for the side that is already taller, the area won't change since the water can still
            # only go up to the shorter wall, which is the same.
            # Thus, we move inwards whichever wall is shorter to try to find a taller wall.
            if left_wall_height < right_wall_height:

                left += 1
            
            else:

                right -=1 
            
        return max_area
